### PR TITLE
ESQL: Don't emit empty pages from hashes

### LINKED ISSUE #112442: [CI] BlockHashTests testBytesRefLongHashHugeCombinatorialExplosion {forcePackedHash=false} failing
**Build Scans:**
- [elasticsearch-periodic-platform-support #3965 / sles-15_platform-support-unix](https://gradle-enterprise.elastic.co/s/qttbqiuurpcws)
- [elasticsearch-periodic-platform-support #3965 / amazonlinux-2_platform-support-aws](https://gradle-enterprise.elastic.co/s/zs7xtbk6wtnfy)
- [elasticsearch-pull-request #30289 / part-3](https://gradle-enterprise.elastic.co/s/aeqtnc3du2nh6)
- [elasticsearch-pull-request #30287 / part-3](https://gradle-enterprise.elastic.co/s/cpxdzrtklih7k)
- [elasticsearch-periodic-platform-support #3917 / debian-11_platform-support-unix](https://gradle-enterprise.elastic.co/s/3aymomkt6tsju)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:esql:compute:test" --tests "org.elasticsearch.compute.aggregation.blockhash.BlockHashTests.testBytesRefLongHashHugeCombinatorialExplosion {forcePackedHash=false}" -Dtests.seed=161991045217ED6D -Dtests.locale=br-Latn-FR -Dtests.timezone=Greenwich -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testBytesRefLongHashHugeCombinatorialExplosion%20%7BforcePackedHash%3Dfalse%7D'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.compute.aggregation.blockhash.BlockHashTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
java.lang.AssertionError: 0: expected not null
```

**Issue Reasons:**
- [main] 5 failures in test testBytesRefLongHashHugeCombinatorialExplosion {forcePackedHash=false} (0.6% fail rate in 843 executions)
- [main] 2 failures in step part-3 (0.8% fail rate in 246 executions)
- [main] 2 failures in pipeline elasticsearch-periodic-platform-support (11.8% fail rate in 17 executions)
- [main] 2 failures in pipeline elasticsearch-pull-request (0.8% fail rate in 244 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

### LINKED ISSUE #112443: [CI] BlockHashTests testBytesRefLongHashHugeCombinatorialExplosion {forcePackedHash=true} failing
**Build Scans:**
- [elasticsearch-periodic-platform-support #3965 / sles-15_platform-support-unix](https://gradle-enterprise.elastic.co/s/qttbqiuurpcws)
- [elasticsearch-periodic-platform-support #3965 / amazonlinux-2_platform-support-aws](https://gradle-enterprise.elastic.co/s/zs7xtbk6wtnfy)
- [elasticsearch-pull-request #30289 / part-3](https://gradle-enterprise.elastic.co/s/aeqtnc3du2nh6)
- [elasticsearch-pull-request #30287 / part-3](https://gradle-enterprise.elastic.co/s/cpxdzrtklih7k)
- [elasticsearch-periodic-platform-support #3917 / debian-11_platform-support-unix](https://gradle-enterprise.elastic.co/s/3aymomkt6tsju)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:esql:compute:test" --tests "org.elasticsearch.compute.aggregation.blockhash.BlockHashTests.testBytesRefLongHashHugeCombinatorialExplosion {forcePackedHash=false}" -Dtests.seed=161991045217ED6D -Dtests.locale=br-Latn-FR -Dtests.timezone=Greenwich -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testBytesRefLongHashHugeCombinatorialExplosion%20%7BforcePackedHash%3Dtrue%7D'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.compute.aggregation.blockhash.BlockHashTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
java.lang.AssertionError: 0: expected not null
```

**Issue Reasons:**
- [main] 5 failures in test testBytesRefLongHashHugeCombinatorialExplosion {forcePackedHash=true} (0.6% fail rate in 843 executions)
- [main] 2 failures in step part-3 (0.8% fail rate in 246 executions)
- [main] 2 failures in pipeline elasticsearch-periodic-platform-support (11.8% fail rate in 17 executions)
- [main] 2 failures in pipeline elasticsearch-pull-request (0.8% fail rate in 244 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

### PR DESCRIPTION
**Pull request by nik9000**
This stops the `BlockHash`es from emitting empty pages when the batch size and number of results from a page lines up. It doesn't hurt the aggs, but it was making testing more complex than we'd like.

Closes #112443
Closes #112442

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BytesRef3BlockHash.java [hunk 1]
```diff
@@ -137,7 +137,7 @@ void add() {
                 }
                 finishMv();
             }
-            emitOrds();
+            flushRemaining();
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BytesRef3BlockHash.java:140]: I could have made this come naturally when you `close()` the thing, but I felt like that'd be confusing.

**Review by alex-spies — APPROVED** [review]
LGTM, thanks Nik!

I noticed the bot muted the failing tests in the meantime. I pushed a change that only unmutes them again, so we don't accidentally close the issues without unmuting.

**Comment by nik9000** [comment]
> I noticed the bot muted the failing tests in the meantime. I pushed a change that only unmutes them again, so we don't accidentally close the issues without unmuting.

Oh thanks!

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/AddBlock.java [hunk 2]
```diff
@@ -135,7 +135,20 @@ protected final void finishMv() {
...
 
-    protected final void emitOrds() {
+    /**
+     * Call when finished to emit all remaining ordinals to the aggs.
+     */
+    protected final void flushRemaining() {
+        if (firstOrd != -1) {
+            throw new IllegalStateException("in the middle of a position");
+        }
+        if (added % emitBatchSize != 0) {
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/AddBlock.java:145]: I think that'd require a the combinatorial explosion of the positions from a block to overflow and int..... That's possible I suppose. I'll just flip it.

FWIW if it takes about 1 nanosecond to process each position then it'd take about two seconds to overflow the int. One nanosecond faster than I'd expect any *grouping* agg to be, but maybe 20 seconds. So, yeah, let's defend against this one.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/AddPage.java [hunk 3]
```diff
@@ -21,13 +21,17 @@
...
+    private final long emitBatchSize;
     private final GroupingAggregatorFunction.AddInput addInput;
 
     private int positionOffset = 0;
-    private int added = 0;
+    /**
+     * Number of added documents. This is a {@code long} because callers will
+     * often perform the combinatorial explosion of values.
+     */
+    private long added = 0;
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/AddPage.java:34]: ++

**Comment by nik9000** [comment]
run buildkite/docs-build-pr

**Comment by nik9000** [comment]
elasticmachine build buildkite/docs-build-pr

**Comment by nik9000** [comment]
Docs build was spurious.

**Pull request merged**